Initial Pallas Matmul support in torch/_inductor.

yarongmu-google opened the pull request:
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

yarongmu-google committed:
Lint.

yarongmu-google committed:
Matmul

yarongmu-google committed:
Lint.

yarongmu-google committed:
Ready for initial feedback.

yarongmu-google commented:
@pytorchbot label "topic: not user facing"

yarongmu-google commented on torch/_inductor/choices.py:
```diff
@@ -162,10 +162,14 @@ def get_ktc(
+                op_name,
+            )
         )
+        print(f"Configs from heuristic for {template_name}: {configs}")
```
Done.

yarongmu-google committed:
Removed debug prints throughput the  PR.

jansel commented on torch/_inductor/codegen/pallas_kernels/block_mm_template.py:
```diff
@@ -0,0 +1,324 @@
+
+    # Move torch tensors to CPU, convert to numpy, then put on TPU device
+    x_tpu_jax = jax.device_put({{x}}.cpu().numpy(), device=jax.devices('tpu')[0])
+    y_tpu_jax = jax.device_put({{y}}.cpu().numpy(), device=jax.devices('tpu')[0])
```
Is there a rush to land this one?  Does anything depend on it?

Once we get the proper plumbing this PR will need to change anyway.

jansel replied:
IMO supporting GPU would improve this PR and not be much code, since it would force the CPU-Tensor hack to be refactored into a common place (rather and copy-and-pasted into the template).

I guess I don't see the point of generating Pallas kernels that do CPU=>TPU=>CPU H2D copies on every kernel call.  It is just creating code we will need to delete later.

Even as a short term enablement thing, I don't see the point.  We can do the same enablement work on GPUs and have the tensors be on the right device.  GPUs let us test correctness better and more easily than using the wrong device.  IMO we should do TPUs right from the start.

jansel replied:
Yes, I am aware they are different -- but part of what PyTorch does is unify many different lower level API to make programming different devices similar.  We need to build abstractions to support both, not just create duplicate code.

yarongmu-google committed:
Fix a minor type casting error.

yarongmu-google committed:
Add the tpu backend to inductor.

yarongmu-google committed:
Enable dlpack for Pallas matmul and add a test workaround.

This commit makes two main changes:

1.  It updates the Pallas `matmul` template to use the `dlpack` protocol for
    zero-copy tensor conversion between PyTorch and JAX. This removes the
    "CPU tensor hack" and will significantly improve performance.

2.  It modifies the `test_matmul` test in `PallasTestsTPU` to use a
    workaround that creates JAX tensors directly on the TPU and converts
    them to PyTorch tensors using `dlpack`. This allows us to test the
    correctness of the `dlpack`-based `matmul` implementation, even
    though we are still blocked by the `torch-tpu` installation issue.

yarongmu-google committed:
Clean up tests.

yarongmu-google committed:
Lint.
